rpc/codec/server.go

```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// Requests need to implement RequestWithUser.
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
```

Comment:
I haven't changed it to a hook yet, just moved the certificate check to connection time (which is the right thing to do either way).
code still depends on security just to check user vs NodeUser (this should probably move).
I'll try to add the hook such that we can use it from the http server as well.